strawberryfields/apps/similarity.py

```diff
@@ -412,119 +527,218 @@ def prob_event_mc(
-    mean_photon_per_mode = n_mean / float(modes)
-
-    p = sf.Program(modes)
-
-    # pylint: disable=expression-not-assigned
-    with p.context as q:
-        sf.ops.GraphEmbed(A, mean_photon_per_mode=mean_photon_per_mode) | q
-
-        if loss:
-            for _q in q:
-                sf.ops.LossChannel(1 - loss) | _q
-
-    eng = sf.LocalEngine(backend="gaussian")
-    result = eng.run(p)
+    State = _get_state(graph, n_mean, loss)
 
     prob = 0
 
-    for _ in range(samples):
+    for _ in range(mc_samples):
         sample = event_to_sample(photon_number, max_count_per_mode, modes)
-        prob += result.state.fock_prob(sample, cutoff=photon_number + 1)
+        prob += State.fock_prob(sample, cutoff=photon_number + 1)
 
-    prob = prob * event_cardinality(photon_number, max_count_per_mode, modes) / samples
+    prob = prob * event_cardinality(photon_number, max_count_per_mode, modes) / mc_samples
 
     return prob
 
 
-def feature_vector_sampling(
-    samples: list, event_photon_numbers: list, max_count_per_mode: int = 2
+def feature_vector_orbits(
+    graph: nx.Graph,
+    list_of_orbits: list,
+    n_mean: float = 5,
+    mc_samples: int = None,
+    loss: float = 0.0,
 ) -> list:
-    r"""Calculates feature vector with respect to input samples.
+    r"""Calculates feature vector made up of either exact or MC-approximated orbit probabilities for the input graph.
 
-    The feature vector is composed of event probabilities with a fixed maximum photon count in
-    each mode but a range of total photon numbers specified by ``event_photon_numbers``.
+    These probabilities can be either exact or estimated using Monte Carlo estimation.
+    The argument ``mc_samples`` is set to ``None`` to get exact feature vector by default.
+    To use Monte Carlo estimation, ``mc_samples`` can be set to the number of samples desired
+    to be used in the estimation.
 
-    Probabilities are reconstructed by measuring the occurrence of events in the input ``samples``.
+    Computing exact probabilities for a large number of orbits especially for orbits with high
+    total photon numbers can be quite time-consuming. For example, calculating the exact
+    probabilities of observing 8 total photons in a 25 mode graph can take on the order of a
+    few minutes. This means that if the dataset has 1000 graphs, obtaining feature vectors for
+    all can take many days. Monte Carlo estimation, although less precise, can be much quicker.
 
     **Example usage:**
 
-    >>> from strawberryfields.apps import data
-    >>> samples = data.Mutag0()
-    >>> feature_vector_sampling(samples, [2, 4, 6])
-    [0.19035, 0.2047, 0.1539]
+    >>> graph = nx.complete_graph(8)
+    >>> feature_vector_orbits(graph, [ [1,1], [2], [2,1,1], [1,1,1,1] ])
+    [0.18423136914879445,
+     7.843627302119571e-31,
+     0.03744399092424445,
+     0.07020748298295838]
+
+    >>> feature_vector_orbits(graph, [ [1,1], [2] , [2,1,1], [1,1,1,1] ], mc_samples = 1000)

```

Comment:
This is very surprising, do the MC estimates really agree so precisely with the exact calculations? I find it hard to believe. I can only spot differences in the 16-th decimal place. No way that's correct

Reply:
I checked the code of ``prob_orbit_exact`` again. I am not sure where the error could be.